wdl-lint/src/rules/disallowed_declaration_name.rs

```diff
@@ -0,0 +1,345 @@
+use wdl_ast::v1::Decl;
+use wdl_ast::v1::PrimitiveTypeKind;
+use wdl_ast::v1::Type;
+use wdl_ast::v1::UnboundDecl;
+
+use crate::Rule;
+use crate::Tag;
+use crate::TagSet;
+
+/// A rule that identifies declaration names that include their type names as a
+/// suffix.
+#[derive(Debug, Default)]
+pub struct DisallowedDeclarationNameRule;
+
+/// Create a diagnostic for a declaration identifier that contains its type name
+fn decl_identifier_with_type(span: Span, decl_name: &str, type_name: &str) -> Diagnostic {
+    Diagnostic::warning(format!(
+        "declaration identifier '{decl_name}' ends with type name '{type_name}'",
+    ))
+    .with_rule("DisallowedDeclarationName")
+    .with_highlight(span)
+    .with_fix("rename the identifier to not include the type name")
+}
+
+impl Rule for DisallowedDeclarationNameRule {
+    fn id(&self) -> &'static str {
+        "DisallowedDeclarationName"
+    }
+
+    fn description(&self) -> &'static str {
+        "Disallows declaration names that end with their type name"
+    }
+
+    fn explanation(&self) -> &'static str {
+        "Declaration names should not include their type as a suffix. This makes the code more \
+         verbose and often redundant. For example, use 'counter' instead of 'counterInt' or \
+         'is_active' instead of 'isActiveBoolean'."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Style, Tag::Clarity])
+    }
+
+    fn exceptable_nodes(&self) -> Option<&'static [SyntaxKind]> {
+        Some(&[
+            SyntaxKind::VersionStatementNode,
+            SyntaxKind::InputSectionNode,
+            SyntaxKind::OutputSectionNode,
+            SyntaxKind::BoundDeclNode,
+            SyntaxKind::UnboundDeclNode,
+        ])
+    }
+}
+
+impl Visitor for DisallowedDeclarationNameRule {
+    type State = Diagnostics;
+
+    fn document(
+        &mut self,
+        _: &mut Self::State,
+        reason: VisitReason,
+        _: &Document,
+        _: SupportedVersion,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // Reset the visitor upon document entry
+        *self = Default::default();
+    }
+
+    fn bound_decl(&mut self, state: &mut Self::State, reason: VisitReason, decl: &BoundDecl) {
+        if reason == VisitReason::Enter {
+            check_decl_name(state, &Decl::Bound(decl.clone()), &self.exceptable_nodes());
+        }
+    }
+
+    fn unbound_decl(&mut self, state: &mut Self::State, reason: VisitReason, decl: &UnboundDecl) {
+        if reason == VisitReason::Enter {
+            check_decl_name(
+                state,
+                &Decl::Unbound(decl.clone()),
+                &self.exceptable_nodes(),
+            );
+        }
+    }
+}
+
+/// Check declaration name for type suffixes
+fn check_decl_name(
+    state: &mut Diagnostics,
+    decl: &Decl,
+    exceptable_nodes: &Option<&'static [SyntaxKind]>,
+) {
+    let name = decl.name();
+    let name_str = name.as_str();
```

Comment:
nit: these don't get used till line 172, and I prefer to keep declarations close to where they're used. Can you move these down? It reduces the need to scroll up and down looking for where something is used/defined.